gax-grpc/src/main/java/com/google/api/gax/grpc/AttemptCallable.java

```diff
@@ -0,0 +1,108 @@
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.grpc;
+
+import com.google.api.core.AbstractApiFuture;
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.api.gax.retrying.RetryingFuture;
+import io.grpc.CallOptions;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+import org.threeten.bp.Duration;
+
+/**
+ * A callable, representing a retriable grpc call. This class is used from {@link RetryingCallable}
+ *
+ * @param <RequestT> request type
+ * @param <ResponseT> response type
+ */
+class AttemptCallable<RequestT, ResponseT> implements Callable<ResponseT> {
+  private final FutureCallable<RequestT, ResponseT> callable;
+  private final RequestT request;
+
+  private volatile RetryingFuture<ResponseT> externalFuture;
+  private volatile CallContext callContext;
+
+  AttemptCallable(
+      FutureCallable<RequestT, ResponseT> callable, RequestT request, CallContext callContext) {
+    this.callable = callable;
+    this.request = request;
+    this.callContext = callContext;
+  }
+
+  public void setExternalFuture(RetryingFuture<ResponseT> externalFuture) {
+    this.externalFuture = externalFuture;
+  }
+
+  @Override
+  public ResponseT call() {
+    if (callContext != null) {
+      callContext =
+          getNextCallContext(callContext, externalFuture.getAttemptSettings().getRpcTimeout());
+    }
+
+    try {
+      externalFuture.setAttemptFuture(new NonCancelableFuture<ResponseT>());
+      if (externalFuture.isDone()) {
+        return null;
+      }
+      ApiFuture<ResponseT> internalFuture = callable.futureCall(request, callContext);
+      externalFuture.setAttemptFuture(internalFuture);
+    } catch (Throwable e) {
+      externalFuture.setAttemptFuture(ApiFutures.<ResponseT>immediateFailedFuture(e));
+      throw e;
+    }
+
+    return null;
+  }
+
+  protected CallContext getNextCallContext(CallContext oldContext, Duration rpcTimeout) {
```

Comment:
Mostly historical reasons. It was essential to an earlier version of the code, but then I changed it. Specifically what now is OperationPollingCallable used to inherit AttemptCallable overriding this method.

After changing it, I still kept the method protected, as it is doing some weird manipulations on call context, which are potentially not desirable by its children (since any non-final class can potentially have children, and automatically agrees to have children, otherwise it would be declared final).

The class itself is package-private, so it and its type hierarchy do not affect surface. I can change it back to private, if it matters.